sdk/src/types/block/output/feature/metadata.rs

```diff
@@ -1,103 +1,250 @@
+    if VERIFY
+        && !MetadataFeature::BYTE_LENGTH_RANGE
+            .contains(&u16::try_from(len).map_err(|e| Error::InvalidMetadataFeature(e.to_string()))?)
+    {
+        return Err(Error::InvalidMetadataFeature(format!(
+            "Out of bounds byte length: {len}"
+        )));
+    }
+    Ok(())
+}
 
 impl MetadataFeature {
     /// The [`Feature`](crate::types::block::output::Feature) kind of [`MetadataFeature`].
     pub const KIND: u8 = 2;
-    /// Valid lengths for a [`MetadataFeature`].
-    pub const LENGTH_RANGE: RangeInclusive<u16> = 1..=8192;
+    /// Valid byte lengths for a [`MetadataFeature`].
+    pub const BYTE_LENGTH_RANGE: RangeInclusive<u16> = 1..=8192;
 
     /// Creates a new [`MetadataFeature`].
     #[inline(always)]
-    pub fn new(data: impl Into<Vec<u8>>) -> Result<Self, Error> {
-        Self::try_from(data.into())
+    pub fn new(data: impl IntoIterator<Item = (String, Vec<u8>)>) -> Result<Self, Error> {
+        let mut builder = Self::build();
+        for (k, v) in data {
+            builder.insert(k, v);
+        }
+        builder.finish()
+    }
+
+    pub fn build() -> MetadataFeatureMap {
+        Default::default()
+    }
+
+    pub fn to_map(&self) -> MetadataFeatureMap {
+        // Unsafe: acceptable because of type level checking, i.e. keys must be ASCII
+        unsafe {
+            MetadataFeatureMap(
+                self.0
+                    .iter()
+                    .map(|(k, v)| {
+                        (
+                            String::from_utf8_unchecked(k.as_ref().to_owned()),
+                            v.as_ref().to_owned(),
+                        )
+                    })
+                    .collect(),
+            )
+        }
     }
 
-    /// Returns the data.
+    /// Returns the data for a given key.
     #[inline(always)]
-    pub fn data(&self) -> &[u8] {
+    pub fn get(&self, key: &str) -> Option<&[u8]> {
+        BoxedSlicePrefix::<u8, MetadataFeatureKeyLength>::try_from(key.as_bytes().to_vec().into_boxed_slice())
+            .ok()
+            .and_then(|key| self.0.get(&key))
+            .map(|v| v.as_ref())
+    }
+}
+
+impl core::ops::Deref for MetadataFeature {
+    type Target = MetadataBTreeMap;
+
+    fn deref(&self) -> &Self::Target {
         &self.0
     }
 }
 
-impl StorageScore for MetadataFeature {}
+/// A map of metadata feature keys to values. This type is not guaranteed to be valid.
+#[derive(Clone, Debug, Default)]
+pub struct MetadataFeatureMap(BTreeMap<String, Vec<u8>>);
 
-impl WorkScore for MetadataFeature {}
+impl MetadataFeatureMap {
+    /// Creates a new [`MetadataFeatureBuilder`].
+    #[inline(always)]
+    pub fn new() -> Self {
+        Self(Default::default())
+    }
 
-macro_rules! impl_from_vec {
-    ($type:ty) => {
-        impl TryFrom<$type> for MetadataFeature {
-            type Error = Error;
+    pub fn with_key_value(mut self, key: &str, value: impl Into<Vec<u8>>) -> Self {
```

Comment:
no warnings for missing docs here (so it's not part of the sdk's API)?

Reply:
We just have #[allow(missing_docs)]

Reply:
Ok was also not exposed